#### Задача 1
Подберите скорость обучения (eta) и количество итераций

In [84]:
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d.axes3d import Axes3D

In [85]:
X = np.array([[ 1,  1],
              [ 1,  1],
              [ 1,  2],
              [ 1,  5],
              [ 1,  3],
              [ 1,  0],
              [ 1,  5],
              [ 1, 10],
              [ 1,  1],
              [ 1,  2]])

y = [45, 55, 50, 55, 60, 35, 75, 80, 50, 60]

In [86]:
# Функция для вычисления MAE 
def calc_mae(y, y_pred):
    err = np.mean(np.abs(y - y_pred))
    return err

# Функция для вычисления MSE 
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

In [87]:
n = X.shape[0] # n=10 - number of objects
 
eta = 0.096 # learning rate
n_iter = 120 # amount of iterations
W = np.array([1, 0.5]) # initial weights

print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):   
    y_pred = np.dot(X, W) # вычисляем y по актуальным весам
    err = calc_mse(y, y_pred) # вычисляем ошибку MSE
    for k in range(W.shape[0]):
        W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    if i % 10 == 0:
        eta /= 1.1 # на каждой десятой итерации снижаем скорость в 1.1 раза
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err, 2)}')

Number of objects = 10        
Learning rate = 0.096        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [11.368 36.692], MSE = 3047.75
Iteration #10: W_new = [ 6566.13481866 36060.78077032], MSE = 5538755658.36
Iteration #20: W_new = [ 2133066.84533434 11762955.62520685], MSE = 787951193917419.2
Iteration #30: W_new = [1.50689618e+08 8.31002977e+08], MSE = 5.340321578247834e+18
Iteration #40: W_new = [2.08852942e+09 1.15175459e+10], MSE = 1.4208490701029318e+21
Iteration #50: W_new = [5.01468492e+09 2.76543218e+10], MSE = 1.163121085479691e+22
Iteration #60: W_new = [1.77593210e+09 9.79367554e+09], MSE = 2.1391163183193936e+21
Iteration #70: W_new = [7.48171553e+07 4.12591524e+08], MSE = 5.811726862876578e+18
Iteration #80: W_new = [ 277664.4777686  1530984.45868739], MSE = 130085535575329.55
Iteration #90: W_new = [102.8567911  323.73725389], MSE = 10096421.16
Iteration #100: W_new = [44.9123317   3.84159772], MSE = 43.98
Iteration #110: W_new = [44.95230315  3.8324825 ], MSE

#### Ответ:
При скорости обучения eta=0.096 и количестве итераций n_iter=110 наибыстрейшим путем достигается значение MSE=43.97, что с учетом округления до 2-го знака после запятой, соответствует расчетному значению MSE, полученному аналитически методом наименьших квадратов

#### Задача 2*
В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее

In [88]:
n = X.shape[0]

eta = 0.054 # learning rate
n_iter = 130

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    #for k in range(W.shape[0]):
    #    W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    # ИЗМЕНЕНИЯ
    W -= eta * (1/n * 2 * (y_pred - y)@X) # поменены местами множители в скалярном произведении
    # ИЗМЕНЕНИЯ
    #
    if i % 10 == 0:
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err,2)}')

Number of objects = 10        
Learning rate = 0.054        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [ 6.832 20.858], MSE = 3047.75
Iteration #10: W_new = [21.50297045 11.41055022], MSE = 582.38
Iteration #20: W_new = [30.68406857  7.5136045 ], MSE = 179.97
Iteration #30: W_new = [36.33244009  5.75528271], MSE = 87.0
Iteration #40: W_new = [39.77660133  4.88931887], MSE = 58.94
Iteration #50: W_new = [41.86677808  4.43090149], MSE = 49.34
Iteration #60: W_new = [43.13201911  4.17535829], MSE = 45.92
Iteration #70: W_new = [43.89684579  4.02808147], MSE = 44.68
Iteration #80: W_new = [44.35882958  3.9414837 ], MSE = 44.23
Iteration #90: W_new = [44.63777106  3.88997318], MSE = 44.06
Iteration #100: W_new = [44.80615588  3.85913373], MSE = 44.0
Iteration #110: W_new = [44.90779016  3.84060347], MSE = 43.98
Iteration #120: W_new = [44.96913087  3.82944724], MSE = 43.97


#### Задача 3*
Вместо того, чтобы задавать количество итераций, задайте другое условие останова алгоритма - когда веса перестают изменяться меньше определенного порога 𝜖

In [89]:
n = X.shape[0] # n=10 - number of objects
 
eta = 0.096 # learning rate
epsilon = 0.0002 # weight change threshold
W = np.array([1, 0.5]) # initial weights

print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')
delta=1
sw0 = np.sum([W])
i=0
while delta >= epsilon:   
    i+=1
    y_pred = np.dot(X, W) # вычисляем y по актуальным весам
    err = calc_mse(y, y_pred) # вычисляем ошибку MSE
    for k in range(W.shape[0]):
        W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    if i % 10 == 0:
        eta /= 1.1 # на каждой десятой итерации снижаем скорость в 1.1 раза
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err, 2)}')
    delta = abs(np.sum(W) - sw0)
    sw0 = np.sum([W])

Number of objects = 10        
Learning rate = 0.096        
Initial weights = [1.  0.5] 

Iteration #10: W_new = [-10993.41218545 -60773.83321208], MSE = 11933914414.63
Iteration #20: W_new = [-15332326.4635356  -84552891.67194663], MSE = 3.046238749953743e+16
Iteration #30: W_new = [-5.00233497e+09 -2.75862164e+10], MSE = 4.33362237784826e+21
Iteration #40: W_new = [-3.53393930e+11 -1.94885017e+12], MSE = 2.9371028656412633e+25
Iteration #50: W_new = [-4.89797384e+12 -2.70106992e+13], MSE = 7.814472994362768e+27
Iteration #60: W_new = [-1.17603303e+13 -6.48543163e+13], MSE = 6.397004793054134e+28
Iteration #70: W_new = [-4.16487739e+12 -2.29679157e+13], MSE = 1.1764843327163981e+28
Iteration #80: W_new = [-1.75459467e+11 -9.67600694e+11], MSE = 3.196369239786435e+25
Iteration #90: W_new = [-6.51067720e+08 -3.59042251e+09], MSE = 7.154524193316631e+20
Iteration #100: W_new = [-135990.49058238 -750186.55125096], MSE = 55528675322321.98
Iteration #110: W_new = [44.22703386 -0.41449177],